File: Zeze/Raft/RocksRaft/LogSet1.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Zeze.Serialize;

namespace Zeze.Raft.RocksRaft
{
	public class LogSet1<V> : LogSet<V>
	{
		public ISet<V> Added { get; } = new HashSet<V>();
		public ISet<V> Removed { get; } = new HashSet<V>();

		public override void Collect(Changes changes, Bean recent, Log vlog)
		{
			throw new Exception($"Collect Not Implement.");
		}

		public bool Add(V item)
		{
			var newset = Value.Add(item);
			if (newset != Value)
            {
				Added.Add(item);
				Removed.Remove(item);
				Value = newset;
				return true;
            }
			return false;
		}

		public bool Remove(V item)
        {
			var newset = Value.Remove(item);
			if (newset != Value)
			{
				Removed.Add(item);
				Added.Remove(item);
				Value = newset;
				return true;
			}
			return false;
        }

		public void Clear()
		{
			foreach (var e in Value)
			{
				Remove(e);
			}
			Value = ImmutableHashSet<V>.Empty;
		}

		public override void Decode(ByteBuffer bb)
		{
			Added.Clear();
			for (int i = bb.ReadInt(); i > 0; --i)
			{
				var value = SerializeHelper<V>.Decode(bb);
				Added.Add(value);
			}

			Removed.Clear();
			for (int i = bb.ReadInt(); i > 0; --i)
			{
				var key = SerializeHelper<V>.Decode(bb);
				Removed.Add(key);
			}
		}

		public override void Encode(ByteBuffer bb)
		{
			bb.WriteInt(Added.Count);
			foreach (var e in Added)
			{
				SerializeHelper<V>.Encode(bb, e);
			}

			bb.WriteInt(Removed.Count);
			foreach (var e in Removed)
			{
				SerializeHelper<V>.Encode(bb, e);
			}
		}

		internal override void EndSavepoint(Savepoint currentsp)
		{
			if (currentsp.Logs.TryGetValue(LogKey, out var log))
			{
				var currentLog = (LogSet1<V>)log;
				currentLog.Value = this.Value;
				currentLog.Merge(this);
			}
			else
			{
				currentsp.Logs[LogKey] = this;
			}
		}

		internal void Merge(LogSet1<V> from)
		{
			// Put,Remove 需要确认有没有顺序问题
			// this: add 1,3 remove 2,4 nest: add 2 remove 1
			foreach (var e in from.Added) Add(e); // replace 1,2,3 remove 4
			foreach (var e in from.Removed) Remove(e); // replace 2,3 remove 1,4
		}

		internal override Log BeginSavepoint()
		{
			var dup = new LogSet<V>();
			dup.Belong = Belong;
			dup.VariableId = VariableId;
			dup.Value = Value;
			return dup;
		}

		public override string ToString()
		{
			var sb = new StringBuilder();
			sb.Append(" Added:");
			ByteBuffer.BuildString(sb, Added);
			sb.Append(" Removed:");
			ByteBuffer.BuildString(sb, Removed);
			return sb.ToString();
		}

	}
}

File: ZezeJava/ZezeJava/src/main/java/Zeze/Transaction/Collections/LogSet1.java
package Zeze.Transaction.Collections;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;
import Zeze.Serialize.ByteBuffer;
import Zeze.Serialize.IByteBuffer;
import Zeze.Transaction.Bean;
import Zeze.Transaction.Log;
import Zeze.Transaction.Savepoint;
import org.jetbrains.annotations.NotNull;
import org.pcollections.Empty;

public class LogSet1<V> extends LogSet<V> {
	protected final @NotNull Meta1<V> meta;
	private final Set<V> added = new HashSet<>();
	private final Set<V> removed = new HashSet<>();

	public LogSet1(Bean belong, int varId, Bean self, @NotNull org.pcollections.PSet<V> value,
				   @NotNull Meta1<V> meta) {
		super(belong, varId, self, value);
		this.meta = meta;
	}

	@Override
	public int getTypeId() {
		return meta.logTypeId;
	}

	@Override
	public @NotNull String getTypeName() {
		return meta.name;
	}

	public final @NotNull Set<V> getAdded() {
		return added;
	}

	public final @NotNull Set<V> getRemoved() {
		return removed;
	}

	public final boolean add(@NotNull V item) {
		var newSet = getValue().plus(item);
		if (newSet != getValue()) {
			removed.remove(item);
			added.add(item);
			setValue(newSet);
			return true;
		}
		return false;
	}

	public final boolean addAll(@NotNull Collection<? extends V> c) {
		var old = getValue();
		// MapPSet.plusAll恒新建包装（same-instance判定失效），逐项plus：单项无变化返回同一
		// 实例，可正确判定"全已存在"（FND7-08）——无变化时不记账不置脏。
		var newSet = old;
		for (V v : c)
			newSet = newSet.plus(v);
		if (newSet == old)
			return false;
		for (V v : c) {
			removed.remove(v);
			added.add(v);
		}
		setValue(newSet);
		return true;
	}

	public final boolean remove(@NotNull V item) {
		var newSet = getValue().minus(item);
		if (newSet != getValue()) {
			added.remove(item);
			removed.add(item);
			setValue(newSet);
			return true;
		}
		return false;
	}

	public final boolean removeAll(@NotNull Collection<? extends V> c) {
		var old = getValue();
		// 同addAll：MapPSet.minusAll恒新建包装，逐项minus判定真实变化（FND7-08）。
		var newSet = old;
		for (V v : c)
			newSet = newSet.minus(v);
		if (newSet == old)
			return false;
		for (V v : c) {
			added.remove(v);
			removed.add(v);
		}
		setValue(newSet);
		return true;
	}

	public final void clear() {
		//for (V v : getValue())
		//	remove(v);
		var exist = getValue();
		added.removeAll(exist);
		removed.addAll(exist);
		setValue(Empty.set());
	}

	@Override
	public void encode(@NotNull ByteBuffer bb) {
		var encoder = meta.valueEncoder;

		bb.WriteUInt(added.size());
		for (V v : added) {
			//noinspection DataFlowIssue
			encoder.accept(bb, v);
		}

		bb.WriteUInt(removed.size());
		for (V v : removed) {
			//noinspection DataFlowIssue
			encoder.accept(bb, v);
		}
	}

	@Override
	public void decode(@NotNull IByteBuffer bb) {
		var decoder = meta.valueDecoder;

		added.clear();
		for (int i = bb.ReadUInt(); i > 0; i--) {
			//noinspection DataFlowIssue
			added.add(decoder.apply(bb));
		}

		removed.clear();
		for (int i = bb.ReadUInt(); i > 0; i--) {
			//noinspection DataFlowIssue
			removed.add(decoder.apply(bb));
		}
	}

	@Override
	public void endSavepoint(@NotNull Savepoint currentSp) {
		Log log = currentSp.getLog(getLogKey());
		if (log != null) {
			@SuppressWarnings("unchecked")
			var currentLog = (LogSet1<V>)log;
			currentLog.setValue(getValue());
			currentLog.merge(this);
		} else
			currentSp.putLog(this);
	}

	private void merge(@NotNull LogSet1<V> from) {
		// add,remove 需要确认有没有顺序问题
		// this: add 1,3 remove 2,4 nest: add 2 remove 1
		for (V v : from.added) {
			removed.remove(v); // replace 1,2,3 remove 4
			added.add(v);
		}
		for (V v : from.removed) {
			added.remove(v); // replace 2,3 remove 1,4
			removed.add(v);
		}
	}

	@Override
	public @NotNull Log beginSavepoint() {
		return new LogSet1<>(getBelong(), getVariableId(), getThis(), getValue(), meta);
	}

	@Override
	public @NotNull String toString() {
		var sb = new StringBuilder();
		sb.append(" added:");
		ByteBuffer.BuildSortedString(sb, added);
		sb.append(" removed:");
		ByteBuffer.BuildSortedString(sb, removed);
		return sb.toString();
	}
}
